### PR TITLE
Feature/osw mod

### LINKED ISSUE #3684: OpenSwath TargetedFileConverter problem with Skyline neutral losses
Hi,

When trying to convert a spectral library generated in Skyline, there seems to be a problem with Neutral losses such as water or ammonia loss.

```
~/Programs/OpenMS/OpenMS_build/bin/TargetedFileConverter -in BD-ZH12Spleen_TransitionList_noprecrsr_sub.tsv -out test.TraML
Error: Unexpected internal error (Could not convert string 'Water Loss (D E S T)' to a double value)
```

the tsv looks like that
```
26      sp|Q9Y5X2|SNX8_HUMAN    SQRFKSSVY       S[Water Loss (D, E, S, T)]Q[Amonia Loss (K, N, Q, R)]R[Amonia Loss (K, N, Q, R)]FK[Amonia Loss (K, N, Q, R)]S[Water Loss (D, E, S, T)]S[Water Loss (D, E, S, T)]VY      14.92436872     551.288005      2       452.237784      2       326875.0313     b       8
30      sp|Q9Y5X2|SNX8_HUMAN    LPADIQAQF       LPAD[Water Loss (D, E, S, T)]IQ[Amonia Loss (K, N, Q, R)]AQ[Amonia Loss (K, N, Q, R)]F  74.44699883     501.76637400000004      2       476.213975      1       1854611.5       y       4
31      sp|Q9Y5X2|SNX8_HUMAN    LPADIQAQF       LPAD[Water Loss (D, E, S, T)]IQ[Amonia Loss (K, N, Q, R)]AQ[Amonia Loss (K, N, Q, R)]F  74.44699883     501.76637400000004      2       510.292225      1       378098.5313     b       5
32      sp|Q9Y5X2|SNX8_HUMAN    LPADIQAQF       LPAD[Water Loss (D, E, S, T)]IQ[Amonia Loss (K, N, Q, R)]AQ[Amonia Loss (K, N, Q, R)]F  74.44699883     501.76637400000004      2       638.3508019999999       1       925691.3125     b       6
33      sp|Q9Y5X2|SNX8_HUMAN    LPADIQAQF       LPAD[Water Loss (D, E, S, T)]IQ[Amonia Loss (K, N, Q, R)]AQ[Amonia Loss (K, N, Q, R)]F  74.44699883     501.76637400000004      2       709.387916      1       461714.0313     b       7
34      sp|Q9Y5X2|SNX8_HUMAN    LPADIQAQF       LPAD[Water Loss (D, E, S, T)]IQ[Amonia Loss (K, N, Q, R)]AQ[Amonia Loss (K, N, Q, R)]F  74.44699883     501.76637400000004      2       837.446494      1       2368778.25      b       8
35      sp|Q9Y5X2|SNX8_HUMAN    LPADIQAQF       LPAD[Water Loss (D, E, S, T)]IQ[Amonia Loss (K, N, Q, R)]AQ[Amonia Loss (K, N, Q, R)]F  74.44699883     501.76637400000004      2       419.226885      2       302268.2813     b       8
39      sp|Q9Y5X2|SNX8_HUMAN    ALKGLSVEF       ALK[Amonia Loss (K, N, Q, R)]GLS[Water Loss (D, E, S, T)]VE[Water Loss (D, E, S, T)]F   68.64774449     482.279117      2       779.4297809999999       1       352151.2813     y       7
40      sp|Q9Y5X2|SNX8_HUMAN    ALKGLSVEF       ALK[Amonia Loss (K, N, Q, R)]GLS[Water Loss (D, E, S, T)]VE[Water Loss (D, E, S, T)]F   68.64774449     482.279117      2       651.3348179999999       1       103822.0938     y       6
41      sp|Q9Y5X2|SNX8_HUMAN    ALKGLSVEF       ALK[Amonia Loss (K, N, Q, R)]GLS[Water Loss (D, E, S, T)]VE[Water Loss (D, E, S, T)]F   68.64774449     482.279117      2       669.429387      1       347736.6875     b       7
42      sp|Q9Y5X2|SNX8_HUMAN    ALKGLSVEF       ALK[Amonia Loss (K, N, Q, R)]GLS[Water Loss (D, E, S, T)]VE[Water Loss (D, E, S, T)]F   68.64774449     482.279117      2       798.47198       1       787750.875      b       8
43      sp|Q9Y5X2|SNX8_HUMAN    ALKGLSVEF       ALK[Amonia Loss (K, N, Q, R)]GLS[Water Loss (D, E, S, T)]VE[Water Loss (D, E, S, T)]F   68.64774449     482.279117      2       780.461415      1       113838.1953     b       8
44      sp|Q9Y5X2|SNX8_HUMAN    ALKGLSVEF       ALK[Amonia Loss (K, N, Q, R)]GLS[Water Loss (D, E, S, T)]VE[Water Loss (D, E, S, T)]F   68.64774449     482.279117      2       399.739628      2       117753.125      b       8
```

### PR DESCRIPTION
**Pull request by hroest**
allow parsing of sequences in OpenSwath that do not map to UniMod

- fixes #3791
- may fix #3684

### CODE REVIEW

File: src/openms/source/ANALYSIS/OPENSWATH/DATAACCESS/DataAccessHelper.cpp [hunk 1]
```diff
@@ -322,12 +322,14 @@ namespace OpenMS
...
 
     aa_sequence = AASequence::fromString(peptide.sequence);
-    for (std::vector<OpenSwath::LightModification>::const_iterator it = peptide.modifications.begin();
-        it != peptide.modifications.end(); ++it)
+    for (const auto & it : peptide.modifications)
     {
-      TargetedExperimentHelper::setModification(it->location, 
-                                                boost::numeric_cast<int>(peptide.sequence.size()), 
-                                                "UniMod:" + String(it->unimod_id), aa_sequence);
+      if (it.unimod_id != -1)
```
Comment (timosachsenberg) [inline src/openms/source/ANALYSIS/OPENSWATH/DATAACCESS/DataAccessHelper.cpp]: add comment or constant?

**Comment by timosachsenberg** [comment]
Thanks for taking care of this issue.
Looks good to me but would be great if @hendrikweisser could also take a quick look to reassure that FFId now properly distinguish PEPTIDE[10.0]R (with unknown modification) and PEPTIDER.
And mixing of unknown and known mods are properly handled? i.e. PM(Oxidation)PTIDE[10.0]R works

**Comment by hendrikweisser** [comment]
> Looks good to me but would be great if @hendrikweisser could also take a quick look to reassure that FFId now properly distinguish PEPTIDE[10.0]R (with unknown modification) and PEPTIDER.

You could easily check that by changing the test file (idXML input) and seeing if the m/z value of the extracted feature gets updated accordingly.

File: src/tests/topp/FeatureFinderIdentification_1_output.featureXML [hunk 2]
```diff
@@ -2490,111 +2490,111 @@
...
 					<overallquality>0.0</overallquality>
 					<charge>0</charge>
-					<UserParam type="float" name="MZ" value="722.826337749971117"/>
-					<UserParam type="float" name="peak_apex_position" value="1788.009033203119998"/>
-					<UserParam type="string" name="native_id" value="YIC(Carbamidomethyl)DNQDTISSK/2_i2"/>
-					<UserParam type="float" name="peak_apex_int" value="1.749065444580078e06"/>
-					<UserParam type="float" name="total_xic" value="1.020859085241699e07"/>
-					<UserParam type="float" name="width_at_50" value="10.2578125"/>
-					<UserParam type="float" name="logSN" value="5.031981761880007"/>
+					<UserParam type="float" name="MZ" value="613.808273177621004"/>
```
Comment (hroest) [inline src/tests/topp/FeatureFinderIdentification_1_output.featureXML]: looks like its almost working, its off by one H+ ion:

```
>>> pyopenms.AASequence.fromString("YIC(Carbamidomethyl)DNQDTISSK").getMonoWeight(pyopenms.Residue.ResidueType.Full, 2) / 2.0
722.3246603310711
>>> pyopenms.AASequence.fromString("YIC[47.021464]D[10.0]NQDTISSK").getMonoWeight(pyopenms.Residue.ResidueType.Full, 2) / 2.0
Warning: unknown modification '47.021464' of residue 'C' - adding it to the database
Warning: unknown modification '10.0' of residue 'D' - adding it to the database
613.306595758721
```

File: src/tests/topp/FeatureFinderIdentification_1_output.featureXML [hunk 3]
```diff
@@ -2490,111 +2490,111 @@
...
 			<UserParam type="float" name="var_yseries_score" value="0.0"/>
-			<UserParam type="float" name="var_dotprod_score" value="0.727520295329778"/>
-			<UserParam type="float" name="var_manhatt_score" value="0.835037543140702"/>
-			<UserParam type="float" name="main_var_xx_swath_prelim_score" value="4.882743484408501"/>
+			<UserParam type="float" name="var_dotprod_score" value="0.23851689113564"/>
+			<UserParam type="float" name="var_manhatt_score" value="1.82859699488381"/>
+			<UserParam type="float" name="main_var_xx_swath_prelim_score" value="3.208144892007533"/>
 			<UserParam type="int" name="missedCleavages" value="0"/>
-			<UserParam type="float" name="PrecursorMZ" value="722.324660331071073"/>
+			<UserParam type="float" name="PrecursorMZ" value="613.30659575872096"/>
```
Comment (hroest) [inline src/tests/topp/FeatureFinderIdentification_1_output.featureXML]: looks like its working correctly:

```
>>> pyopenms.AASequence.fromString("YIC(Carbamidomethyl)DNQDTISSK").getMonoWeight(pyopenms.Residue.ResidueType.Full, 2) / 2.0
722.3246603310711
>>> pyopenms.AASequence.fromString("YIC[47.021464]D[10.0]NQDTISSK").getMonoWeight(pyopenms.Residue.ResidueType.Full, 2) / 2.0
Warning: unknown modification '47.021464' of residue 'C' - adding it to the database
Warning: unknown modification '10.0' of residue 'D' - adding it to the database
613.306595758721
```

**Comment by hroest** [comment]
> You could easily check that by changing the test file (idXML input) and seeing if the m/z value of the extracted feature gets updated accordingly.

the m/z file is still something that is computed in FFId - OpenSWATH will only parse the AASequence for the b/y ion series but it will work with the provided precursor m/z which is calculated outside OpenSWATH. 

> And mixing of unknown and known mods are properly handled? i.e. PM(Oxidation)PTIDE[10.0]R works

should work as long as AASequence can parse it

**Pull request merged**